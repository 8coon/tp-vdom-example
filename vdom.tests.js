const vdom = require('./vdom');

describe('vdom', () => {

	describe('create', () => {
		test('Создаёт одну текстовую ноду', () => {
			expect(vdom.create('It works!')).toMatchSnapshot();
		});

		test('Создаёт один элемент', () => {
			expect(vdom.create({tag: 'DIV'})).toMatchSnapshot();
		});

		test('Создаёт один элемент с двумя атрибутами', () => {
			expect(vdom.create({
				tag: 'DIV',
				attrs: {
					style: 'background-color: red;',
					'tab-index': 0
				}
			})).toMatchSnapshot();
		});

		test('Создаёт один элемент с двумя детьми', () => {
			expect(vdom.create({
				tag: 'DIV',
				children: [
					{tag: 'SPAN'},
					'Some text'
				]
			})).toMatchSnapshot();
		});
	});

	describe('update', () => {
		describe('attrs', () => {
			test('Обновляет атрибуты 2 -> 2', () => {
				const node = vdom.create({
					tag: 'DIV',
					attrs: {
						style: 'background-color: red;',
						'tab-index': 0
					}
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					attrs: {
						style: 'background-color: black;',
						'tab-index': 2
					}
				})).toMatchSnapshot();
			});

			test('Обновляет атрибуты 2 -> 1', () => {
				const node = vdom.create({
					tag: 'DIV',
					attrs: {
						style: 'background-color: red;',
						'tab-index': 0
					}
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					attrs: {
						style: 'background-color: black;'
					}
				})).toMatchSnapshot();
			});

			test('Обновляет атрибуты 1 -> 2', () => {
				const node = vdom.create({
					tag: 'DIV',
					attrs: {
						'tab-index': 0
					}
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					attrs: {
						style: 'background-color: black;',
						'tab-index': 2
					}
				})).toMatchSnapshot();
			});

			test('Обновляет атрибуты 2 -> 0', () => {
				const node = vdom.create({
					tag: 'DIV',
					attrs: {
						style: 'background-color: red;',
						'tab-index': 0
					}
				});

				expect(vdom.update(node, {
					tag: 'DIV',
				})).toMatchSnapshot();
			});

			test('Обновляет атрибуты 0 -> 2', () => {
				const node = vdom.create({
					tag: 'DIV',
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					attrs: {
						style: 'background-color: black;',
						'tab-index': 2
					}
				})).toMatchSnapshot();
			});
		});

		describe('children', () => {
			test('Обновляет детей 1 -> 1', () => {
				const node = vdom.create({
					tag: 'DIV',
					children: [
						{tag: 'SPAN', key: 'A'},
					]
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					children: [
						{tag: 'SPAN', key: 'A'},
					]
				})).toMatchSnapshot();
			});

			test('Обновляет детей 1 -> 1 разные типы', () => {
				const node = vdom.create({
					tag: 'DIV',
					children: [
						{tag: 'SPAN', key: 'A'},
					]
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					children: [
						'Some text',
					]
				})).toMatchSnapshot();
			});

			test('Обновляет детей 1 -> 1 разные ключи', () => {
				const node = vdom.create({
					tag: 'DIV',
					children: [
						{tag: 'SPAN', key: 'A'},
					]
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					children: [
						{tag: 'B', key: 'B'},
					]
				})).toMatchSnapshot();
			});

			test('Обновляет детей 2 -> 2', () => {
				const node = vdom.create({
					tag: 'DIV',
					children: [
						{tag: 'SPAN', key: 'A'},
						'Some text'
					]
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					children: [
						{tag: 'SPAN', attrs: {disabled: 'disabled'}, key: 'A'},
						'Some text updated'
					]
				})).toMatchSnapshot();
			});

			test('Обновляет детей 2 -> 2 разные типы', () => {
				const node = vdom.create({
					tag: 'DIV',
					children: [
						{tag: 'SPAN', key: 'A'},
						'Some text'
					]
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					children: [
						'Some text updated',
						{tag: 'SPAN', attrs: {disabled: 'disabled'}, key: 'A'},
					]
				})).toMatchSnapshot();
			});

			test('Обновляет детей 2 -> 2 разные ключи', () => {
				const node = vdom.create({
					tag: 'DIV',
					children: [
						{tag: 'B', attrs: {A: 'A'}, key: 'A'},
						{tag: 'I', attrs: {B: 'B'}, key: 'B'},
					]
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					children: [
						{tag: 'B', attrs: {B: 'B'}, key: 'C'},
						{tag: 'I', attrs: {A: 'A'}, key: 'D'},
					]
				})).toMatchSnapshot();
			});

			test('Обновляет детей 2 -> 0', () => {
				const node = vdom.create({
					tag: 'DIV',
					children: [
						{tag: 'B', attrs: {A: 'A'}, key: 'A'},
						{tag: 'I', attrs: {B: 'B'}, key: 'B'},
					]
				});

				expect(vdom.update(node, {
					tag: 'DIV'
				})).toMatchSnapshot();
			});

			test('Обновляет детей 0 -> 2', () => {
				const node = vdom.create({
					tag: 'DIV'
				});

				expect(vdom.update(node, {
					tag: 'DIV',
					children: [
						{tag: 'B', attrs: {A: 'A'}, key: 'A'},
						{tag: 'I', attrs: {B: 'B'}, key: 'B'},
					]
				})).toMatchSnapshot();
			});
		});
	});

	describe('components', () => {
		function createLoggingComponentClass(log = [], prefix = '') {

			class LoggingComponent extends vdom.Component {
				didCreate() {
					super.didCreate();
					log.push(prefix + 'didCreate');
				}
				willUpdate(attrs, children) {
					super.willUpdate(attrs, children);
					log.push(prefix + 'willUpdate');
				}
				didUpdate() {
					super.didUpdate();
					log.push(prefix + 'didUpdate');
				}
				willDestroy() {
					super.willDestroy();
					log.push(prefix + 'willDestroy');
				}
			}

			return [LoggingComponent, log];
		}

		test('Отображает компонент', () => {
			const [LoggingComponent, log] = createLoggingComponentClass();

			class Test extends LoggingComponent {
				render() {
					return {tag: 'div'}
				}
			}

			const node = vdom.create({tag: Test});
			expect(node).toMatchSnapshot();

			vdom.destroy(node);
			expect(log).toMatchSnapshot();
		});

		test('Обновляет компонент с атрибутом', () => {
			const [LoggingComponent, log] = createLoggingComponentClass();

			class Test extends LoggingComponent {
				render() {
					return {
						tag: 'div',
						children: [this.attrs.text]
					}
				}
			}

			const node = vdom.create({tag: Test, attrs: {text: 'Some text'}});
			expect(node).toMatchSnapshot();

			vdom.update(node, {attrs: {text: 'Some text updated'}});
			expect(node).toMatchSnapshot();

			vdom.destroy(node);
			expect(log).toMatchSnapshot();
		});

		test('Обновляет вложенные компоненты', () => {
			const [LoggingParent, log] = createLoggingComponentClass();
			const [LoggingChild] = createLoggingComponentClass(log, 'child ');

			class Child extends LoggingChild {
				render() {
					return {
						tag: 'span',
						children: [this.attrs.text]
					}
				}
			}

			class Parent extends LoggingParent {
				render() {
					return {
						tag: 'div',
						children: [{
							tag: Child,
							attrs: {text: `Enough of ${this.attrs.text}`},
							key: `child-key-${this.attrs.key}`
						}]
					}
				}
			}

			const node = vdom.create({
				tag: Parent,
				attrs: {text: 'Some text', key: '1'}
			});
			expect(node).toMatchSnapshot();

			vdom.update(node, {
				tag: Parent,
				attrs: {text: 'Some text updated', key: '1'}
			});
			expect(node).toMatchSnapshot();

			vdom.update(node, {
				tag: Parent,
				attrs: {text: 'Some text other key', key: '2'}
			});
			expect(node).toMatchSnapshot();

			vdom.destroy(node);
			expect(log).toMatchSnapshot();
		});
	});

});
