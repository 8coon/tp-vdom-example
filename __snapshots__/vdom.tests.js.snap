// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vdom components Обновляет вложенные компоненты 1`] = `
<div>
  <span>
    Enough of Some text
  </span>
</div>
`;

exports[`vdom components Обновляет вложенные компоненты 2`] = `
<div>
  <span>
    Enough of Some text updated
  </span>
</div>
`;

exports[`vdom components Обновляет вложенные компоненты 3`] = `
<div>
  <span>
    Enough of Some text other key
  </span>
</div>
`;

exports[`vdom components Обновляет вложенные компоненты 4`] = `
Array [
  "child didCreate",
  "didCreate",
  "willUpdate",
  "child willUpdate",
  "child didUpdate",
  "didUpdate",
  "willUpdate",
  "child didCreate",
  "child willDestroy",
  "didUpdate",
  "willDestroy",
  "child willDestroy",
]
`;

exports[`vdom components Обновляет компонент с атрибутом 1`] = `
<div>
  Some text
</div>
`;

exports[`vdom components Обновляет компонент с атрибутом 2`] = `
<div>
  Some text updated
</div>
`;

exports[`vdom components Обновляет компонент с атрибутом 3`] = `
Array [
  "didCreate",
  "willUpdate",
  "didUpdate",
]
`;

exports[`vdom components Отображает компонент 1`] = `<div />`;

exports[`vdom components Отображает компонент 2`] = `
Array [
  "didCreate",
  "willDestroy",
]
`;

exports[`vdom create Создаёт один элемент 1`] = `<div />`;

exports[`vdom create Создаёт один элемент с двумя атрибутами 1`] = `
<div
  style="background-color: red;"
  tab-index="0"
/>
`;

exports[`vdom create Создаёт один элемент с двумя детьми 1`] = `
<div>
  <span />
  Some text
</div>
`;

exports[`vdom create Создаёт одну текстовую ноду 1`] = `It works!`;

exports[`vdom update attrs Обновляет атрибуты 0 -> 2 1`] = `
<div
  style="background-color: black;"
  tab-index="2"
/>
`;

exports[`vdom update attrs Обновляет атрибуты 1 -> 2 1`] = `
<div
  style="background-color: black;"
  tab-index="2"
/>
`;

exports[`vdom update attrs Обновляет атрибуты 2 -> 0 1`] = `<div />`;

exports[`vdom update attrs Обновляет атрибуты 2 -> 1 1`] = `
<div
  style="background-color: black;"
/>
`;

exports[`vdom update attrs Обновляет атрибуты 2 -> 2 1`] = `
<div
  style="background-color: black;"
  tab-index="2"
/>
`;

exports[`vdom update children Обновляет детей 0 -> 2 1`] = `
<div>
  <b
    a="A"
  />
  <i
    b="B"
  />
</div>
`;

exports[`vdom update children Обновляет детей 1 -> 1 1`] = `
<div>
  <span />
</div>
`;

exports[`vdom update children Обновляет детей 1 -> 1 разные ключи 1`] = `
<div>
  <b />
</div>
`;

exports[`vdom update children Обновляет детей 1 -> 1 разные типы 1`] = `
<div>
  Some text
</div>
`;

exports[`vdom update children Обновляет детей 2 -> 0 1`] = `<div />`;

exports[`vdom update children Обновляет детей 2 -> 2 1`] = `
<div>
  <span
    disabled="disabled"
  />
  Some text updated
</div>
`;

exports[`vdom update children Обновляет детей 2 -> 2 разные ключи 1`] = `
<div>
  <b
    b="B"
  />
  <i
    a="A"
  />
</div>
`;

exports[`vdom update children Обновляет детей 2 -> 2 разные типы 1`] = `
<div>
  Some text updated
  <span
    disabled="disabled"
  />
</div>
`;
